src/host/getset.cpp

```diff
@@ -1028,36 +1036,89 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
                 return S_OK;
             }
 
-            const auto clipViewport = clip ? Viewport::FromInclusive(*clip) : bufferSize;
+            const auto clipViewport = clip ? Viewport::FromInclusive(*clip).Clamp(bufferSize) : bufferSize;
             const auto sourceViewport = Viewport::FromInclusive(source);
-            Viewport readViewport;
-            Viewport writtenViewport;
+            const auto fillViewport = sourceViewport.Clamp(clipViewport);
 
-            const auto w = std::max(0, sourceViewport.Width());
-            const auto h = std::max(0, sourceViewport.Height());
-            const auto a = static_cast<size_t>(w * h);
-            if (a == 0)
+            writer.BackupCursor();
+
+            if (gci.GetVtIo()->GetDeviceAttributes().test(Microsoft::Console::VirtualTerminal::DeviceAttribute::RectangularAreaOperations))
             {
-                return S_OK;
+                // This calculates just the positive offsets caused by out-of-bounds source and target coordinates.
+                // The source gets offset because it's outside the bounds of the buffer (= negative),
+                // while the target gets offset because it's outside the bounds of the clip rect (= usually negative).
+                const til::point offset{
+                    std::max(0, -source.left) + std::max(0, clipViewport.Left() - target.x),
+                    std::max(0, -source.top) + std::max(0, clipViewport.Top() - target.y),
+                };
+
+                const auto copyTargetViewport = Viewport::FromDimensions(target + offset, sourceViewport.Dimensions()).Clamp(clipViewport);
+                const auto copySourceViewport = Viewport::FromDimensions(sourceViewport.Origin() + offset, copyTargetViewport.Dimensions()).Clamp(bufferSize);
+                const auto fills = Viewport::Subtract(fillViewport, copyTargetViewport);
+                std::wstring buf;
+
+                if (!fills.empty())
+                {
+                    Microsoft::Console::VirtualTerminal::VtIo::FormatAttributes(buf, TextAttribute{ fillAttribute });
+                }
+
+                if (copySourceViewport.IsValid() && copyTargetViewport.IsValid())

```

Comment:
If we bail out here, is there still something we could have done by manual-copy? e.g. if it's not representable with DECCRA, is there still a potential fallback?

Reply:
No, this just means that there was nothing to copy because either the source or target rects are completely out of bounds and so there's nothing to copy.